benchmarks/graphviz/graphviz-2.34.0/cmd/tools/mm2gv.1
.EQ
delim $$
.EN
.TH MM2GV 1 "31 July 2008"
.SH NAME
mm2gv \- Matrix Market-DOT converters
.SH SYNOPSIS
.B mm2gv
[
.B \-cluv?
]
[
.BI -U i
]
[
.BI -o outfile
]
[ 
.I file
]
.br
.SH DESCRIPTION
.B mm2gv
converts a sparse matrix of the Matrix Market format to a graph in the GV (formerly DOT) format. 
If the matrix \fIM\fP is not square, the graph is considered bipartite and the matrix is viewed
as a bipartite graph adjacency matrix, with the rows and columns of the matrix specifying the
two sets of vertices. Equivalently, the matrix is converted into a symmetric square matrix
.TS
c c.
0	\fIM\fP
$\fIM\fP sup \fIT\fP$	0
.TE
a block matrix with square blocks of 0's in the upper left and lower right, the upper right
block being \fIM\fP and the lower left block being the transpose of \fIM\fP.
This matrix is then viewed as the adjacency matrix of the graph.
.P
For a square matrix, \fBmm2gv\fP uses it directly as an adjacency matrix if its pattern of non-zero
entries is symmetric; otherwise, it will treat it as a bipartite graph as with the case of non-square
matrices. This behavior can be modified by the \fB-U\fP flag.
.SH OPTIONS
The following options are supported:
.TP
.B \-c
This flag causes \fBmm2gv\fP to assign colors to the edges. The matrix element is scaled to the
range [0,1] depending on where it lies between the minimum and maximum set matrix values. This
scaled value is used as the \fI"wt"\fP attribute of the corresponding edge. 
In addition, this scalar value is mapped to an RGB value, which is stored as the edge \fI"color"\fP.
.TP
.B \-l
If set, \fBmm2gv\fP attaches a label to the graph indicating the base name of the input file,
and the number of nodes and edges.
.TP
.B \-u
If specified, the graph is assumed to be undirected. By default, the graph generated is directed. 
.TP
.B \-v
This flag causes \fBmm2gv\fP to store the matrix values as the \fI"len"\fP attribute of the
corresponding edge.
.TP
.BI \-U "bflag"
Specifies how square matrices are handled. If \fIbflag\fP is 0, a square matrix will always be treated
as an adjacency matrix.  
If \fIbflag\fP is 1 (the default), a square matrix with a symmetric pattern
of non-zero entries will be used as an adjacency matrix; otherwise, it will be used a bipartite graph.
If \fIbflag\fP is 2, a symmetric matrix
will be used as an adjacency matrix; otherwise, it will be used a bipartite graph.
If \fIbflag\fP is 3, any input matrix will be treated like a bipartite graph.
.TP
.BI \-o "outfile"
Prints output to the file \fIoutfile\fP. If not given, \fBmm2gv\fP
uses stdout.
.TP
.SH OPERANDS
The following operand is supported:
.TP 8
.I file
Name of the file in MatrixMarket format.
If no
.I file
operand is specified,
the standard input will be used.
.SH RETURN CODES
Return \fB0\fP
if there were no problems during conversion;
and non-zero if any error occurred.
.SH AUTHORS
Yifan Hu <yifanhu@research.att.com>
.br
Emden R. Gansner <erg@research.att.com>
.SH ADDITIONAL INFO
See http://math.nist.gov/MatrixMarket/ for description of the format and http://www.cise.ufl.edu/research/sparse/matrices/ for a large collection of sparse matrices in this format.


NEWS
KLEE 1.4.0, 21 July 2017
========================
(Incorporating changes from 4 November 2016 up to and including 21 July 2017)
Documentation at http://klee.github.io/releases/docs/v1.4.0/

This will be the last version supporting LLVM 2.9 and the autoconf build system.

- New CMake build system (@delcypher, @jirislaby)
- Added support for vectorized instructions (@delcypher)
- Fixed and documents BFS searcher behaviour (@MartinNowack, @ccadar)
- Renames .pc files to .kquery files (@holycrap872)
- Removed option --randomize-fork (@ccadar)
- Changed preferred permissions from 0622 to the more standard 0644 in the
POSIX model (@ccadar)
- New target name, "make systemtests", for running the system tests.  This replaces "make test".  Running the unit tests is still accomplished via "make unittests".
- Better support for MacOS (@andreamattavelli, @delcypher)
- Enabled support for ASan builds of KLEE (@delcypher)
- Support long long values in --stop-after-n-instructions for LLVM > 2.9
(@andreamattavelli)
- Teach KLEE to respect the requested memory alignment of globals and stack
variables when possible (@delcypher)
- Added new option --warnings-only-to-file which causes warnings to be written
to warnings.txt only (@ccadar)
- metaSMT improvements (@hoangmle)
- KLEE-web improvements (@andronat, @helicopter88)
- Fixed bug in the implementation of NotExpr (@delcypher)
- Fixed a bug leading to data loss when writing into files (@ccadar, @gladtbx)
- Some improvements and refactoring to the Expr library (@delcypher)
- Added missing constant folding opportunity when handling constant arrays
(@andreamattavelli, @delcypher)
- Teach klee::getDirectCallTarget() to resolve weak aliases (@delcypher)
- Fixed handling of internal forks (@gladtbx)
- Improved replay using libkleeRuntest (@delcypher)
- Added -debug-assignment-validating-solver feature to check the correctness
of generated assignments (@delcypher)
- Added -debug-z3-dump-queries, -debug-z3-validate-models and
-debug-z3-verbosity options useful for debugging the interaction with Z3
(@delcypher)
- Added geq/lt-llvm- configs in lit (@jirislaby)
- Work on supporting newer LLVM versions (@jirislaby)
- Fixed bug where stats would not be updated on early exit (@delcypher)
- Reworked the external dispatching mechanism (@delcypher)
- Added support for creating release documentation (@delcypher)
- Smaller refactorings, fixes and improvements, test cases, maintenance,
comments, website, etc. (@adrianherrera, @akshaynagpal, @AlexxNica,
@andreamattavelli, @bananaappletw, @bigelephant29, @bunseokbot, @ccadar,
@delcypher, @emlai, @hoangmle, @jirislaby, @kren1, @levex,
@Manouchehri, @MartinNowack, @mechtaev, @Mic92, @omeranson, @rtrembecky,
@thestr4ng3r, @tomek-kuchta)


KLEE 1.3.0, 30 November 2016
============================
(Incorporating changes from 1 April up to and including 3 November 2016)

* Improved determinism of KLEE, an essential feature for experiments involving KLEE (@MartinNowack)
* KLEE-web has been improved and refactored, and now available at http://klee.doc.ic.ac.uk/ (@giacomoguerci, @helicopter88, @andronat, @ccadar, based on work by @ains, @ben-chin, @ilovepjs, @JamesDavidCarr, Kaho Sato, Conrad Watt, @ccadar)
* Renamed --replay-out to --replay-ktest and --replay-out-dir to replay-ktest-dir (@delcypher)
* Split creation of symbolic files and stdin in two distinct options, documented at http://klee.github.io/docs/options/#symbolic-environment (@andreamattavelli)
* Support for logging queries before invoking the solver via --log-partial-queries-early, useful for debugging solver crashes (@MartinNowack)
* Added --stats-write-after-instructions and --istats-write-after-instructions to update each statistic after n steps (@MartinNowack)
* Added --compress-log and --debug-compress-instructions to gzip-compress logs (@MartinNowack)
* Added --exit-on-error-type option for stopping execution when certain error types are encountered (@jirislaby)
* Updated and improved metaSMT support and added TravisCI targets (@hoangmle)
* Added option --debug-crosscheck-core-solver to allow crosschecking of solvers (@delcypher)
* Explicitly made division total in STP (@ccadar)
* Extended support for assembler raising (@MartinNowack)
* Disabled --solver-optimize-divides, as the optimization is currently buggy (@ccadar)
* Improved --debug-print-instructions options with more logging options (@andreamattavelli)
* Improved stub for times() not to trigger a NULL dereference (@ccadar)
* Allow relocation of installed KLEE tree (@ShayDamir)
* Fixed bug in independent solver (@delcypher)
* Fixed alignement of varargs (@MartinNowack)
* Fixed variable shifting behavior with different sizes and generation of STP shift operations with variable amounts (@MartinNowack)
* Fixed handling of non-sized globals (@jirislaby)
* Fixed klee_get_obj_size() crash on 64-bit (@hutoTUM)
* Fixed bug in Kleaver's parser (@andreamattavelli)
* Refactorings, small fixes and improvements, test cases, maintenance and website: (@andreamattavelli, @ccadar, @delcypher, @domainexpert, @giacomoguerci, @hoangmle, @helicopter88, @jirislaby, @Justme0, @kren1, @MartinNowack, @mchalupa)


KLEE 1.2.0, 31 March 2016
=========================

* Added native support for Z3 (@delcypher)
* Made it possible to build KLEE without using STP and only MetaSMT (@delcypher)
* Added support for tcmalloc, which allows KLEE to better track memory consumption (@MartinNowack)
* Added support for lowering the ``llvm.objectsize`` intrinsic (@delcypher)
* Added soname for Runtest dynamic library (@MartinNowack)
* Added support to load libraries from command line (@omeranson)
* Added command line flag --silent-klee-assume to suppress errors due to infeasible assumptions (Valentin Wüstholz, @wuestholz)
* Changed code to print out arrays deterministically (@MartinNowack)
* Improved klee-clang script (@msoos)
* Added code to dump queries and assignments (@MartinNowack)
* Code cleanup and refactorings (@delcypher, @MartinNowack)

benchmarks/graphviz/graphviz-2.34.0/cmd/tools/mm2gv.c
/* $Id$Revision:  */
/* vim:set shiftwidth=4 ts=8: */

/*************************************************************************
 * Copyright (c) 2011 AT&T Intellectual Property 
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors: See CVS logs. Details at http://www.graphviz.org/
 *************************************************************************/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#define STANDALONE
#include "cgraph.h"
/* #include "arith.h" */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <assert.h>
#include "mmio.h"
#include "agxbuf.h"
#include "SparseMatrix.h"
#include "matrix_market.h"
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#else
#include "compat_getopt.h"
#endif

#define MALLOC malloc
#define FREE free
#define test_flag(a, flag) ((a)&(flag))
#define real double
#define BUFS         1024

#ifdef WIN32 //*dependencies
    #pragma comment( lib, "cgraph.lib" )
    #pragma comment( lib, "ingraphs.lib" )
    #pragma comment( lib, "gvc.lib" )
    #pragma comment( lib, "sparse.lib" )
#endif


typedef struct {
    Agrec_t h;
    int id;
} Agnodeinfo_t;

#define ND_id(n)  (((Agnodeinfo_t*)(n->base.data))->id)

static char *cmd;

static real Hue2RGB(real v1, real v2, real H)
{
    if (H < 0.0)
	H += 1.0;
    if (H > 1.0)
	H -= 1.0;
    if ((6.0 * H) < 1.0)
	return (v1 + (v2 - v1) * 6.0 * H);
    if ((2.0 * H) < 1.0)
	return v2;
    if ((3.0 * H) < 2.0)
	return (v1 + (v2 - v1) * ((2.0 / 3.0) - H) * 6.0);
    return v1;
}

char *hex[16] =
    { "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "a", "b", "c", "d",
"e", "f" };

static char *hue2rgb(real hue, char *color)
{
    real v1, v2, lightness = .5, saturation = 1;
    int red, blue, green;

    if (lightness < 0.5)
	v2 = lightness * (1.0 + saturation);
    else
	v2 = (lightness + saturation) - (saturation * lightness);

    v1 = 2.0 * lightness - v2;

    red = (int) (255.0 * Hue2RGB(v1, v2, hue + (1.0 / 3.0)) + 0.5);
    green = (int) (255.0 * Hue2RGB(v1, v2, hue) + 0.5);
    blue = (int) (255.0 * Hue2RGB(v1, v2, hue - (1.0 / 3.0)) + 0.5);
    color[0] = '#';
    sprintf(color + 1, "%s", hex[red / 16]);
    sprintf(color + 2, "%s", hex[red % 16]);
    sprintf(color + 3, "%s", hex[green / 16]);
    sprintf(color + 4, "%s", hex[green % 16]);
    sprintf(color + 5, "%s", hex[blue / 16]);
    sprintf(color + 6, "%s", hex[blue % 16]);
    color[7] = '\0';
    return color;
}

#if 0
static void posStr(char *buf, int dim, real * x, double sc)
{
    if (dim == 3) {
	sprintf(buf, "%f,%f,%f", sc * x[0], sc * x[1], sc * x[2]);
    } else {
	sprintf(buf, "%f,%f", sc * x[0], sc * x[1]);
    }
}

static void attach_embedding(Agraph_t * g, int dim, double sc, real * x)
{
    Agsym_t *sym = agattr(g, AGNODE, "pos", 0);
    Agnode_t *n;
    char buf[BUFS];
    int i = 0;

    if (!sym)
	sym = agattr(g, AGNODE, "pos", "");

    for (n = agfstnode(g); n; n = agnxtnode(g, n)) {
	assert(i == ND_id(n));
	posStr(buf, dim, x + i * dim, sc);
	agxset(n, sym, buf);
	i++;
    }

}

/* SparseMatrix_import_dot:
 * Assumes g is connected and simple, i.e., we can have a->b and b->a
 * but not a->b and a->b
 */
SparseMatrix
SparseMatrix_import_dot(Agraph_t * g, int dim, real ** label_sizes,
			real ** x, int format)
{
    SparseMatrix A = 0;
    Agnode_t *n;
    Agedge_t *e;
    Agsym_t *sym;
    int nnodes;
    int nedges;
    int i, row;
    int *I;
    int *J;
    real *val;
    real v;
    int type = MATRIX_TYPE_REAL;
    real padding = 10;

    if (!g)
	return NULL;
    nnodes = agnnodes(g);
    nedges = agnedges(g);
    if (format != FORMAT_CSR) {
	fprintf(stderr, "Format %d not supported\n", format);
	exit(1);
    }

    /* Assign node ids */
    i = 0;
    for (n = agfstnode(g); n; n = agnxtnode(g, n))
	ND_id(n) = i++;

    I = N_NEW(nedges, int);
    J = N_NEW(nedges, int);
    val = N_NEW(nedges, real);

    sym = agndattr(g->proto->e, "wt");
    i = 0;
    for (n = agfstnode(g); n; n = agnxtnode(g, n)) {
	row = ND_id(n);
	for (e = agfstout(g, n); e; e = agnxtout(g, e)) {
	    I[i] = row;
	    J[i] = ND_id(e->head);
	    if (sym) {
		if (sscanf(agxget(e, sym->index), "%lf", &v) != 1)
		    v = 1;
	    } else
		v = 1;
	    val[i] = v;
	    i++;
	}
    }

    *label_sizes = N_NEW(2 * nnodes, real);
    for (n = agfstnode(g); n; n = agnxtnode(g, n)) {
	real sz;
	i = ND_id(n);
	if (agget(n, "width") && agget(n, "height")) {
	    sscanf(agget(n, "width"), "%lf", &sz);
	    /*      (*label_sizes)[i*2] = POINTS(sz)*.6; */
	    (*label_sizes)[i * 2] = POINTS(sz) * .5 + padding * 0.5;
	    sscanf(agget(n, "height"), "%lf", &sz);
	    /*(*label_sizes)[i*2+1] = POINTS(sz)*.6; */
	    (*label_sizes)[i * 2 + 1] = POINTS(sz) * .5 + padding * 0.5;
	} else {
	    (*label_sizes)[i * 2] = 4 * POINTS(0.75) * .5;
	    (*label_sizes)[i * 2 + 1] = 4 * POINTS(0.5) * .5;
	}
    }

    if (x) {
	*x = N_NEW(dim * nnodes, real);
	for (n = agfstnode(g); n; n = agnxtnode(g, n)) {
	    real xx, yy;
	    i = ND_id(n);
	    if (agget(n, "pos")) {
		sscanf(agget(n, "pos"), "%lf,%lf", &xx, &yy);
		(*x)[i * dim] = xx;
		(*x)[i * dim + 1] = yy;
	    } else {
		(*x)[i * dim] = 0;
		(*x)[i * dim + 1] = 0;
	    }
	}
    }

    A = SparseMatrix_from_coordinate_arrays(nedges, nnodes, nnodes, I, J,
					    val, type);

    FREE(I);
    FREE(J);
    FREE(val);

    return A;
}
#endif

static Agraph_t *makeDotGraph(SparseMatrix A, char *name, int dim,
			      real * x, int with_color, int with_label, int with_val)
{
    Agraph_t *g;
    Agnode_t *n;
    Agnode_t *h;
    Agedge_t *e;
    int i, j;
    agxbuf xb;
    char buf[BUFS];
    unsigned char string[BUFS];
    Agsym_t *sym, *sym2 = NULL, *sym3 = NULL;
    int *ia = A->ia;
    int *ja = A->ja;
    real *val = (real *) (A->a);
    Agnode_t **arr = N_NEW(A->m, Agnode_t *);
    real *color = NULL;
    char cstring[8];

    name = strip_dir(name);

    if (with_color) {
	if ((A->type == MATRIX_TYPE_REAL) && !val) {
	    fprintf (stderr, "Warning: input matrix has no values, -c flag ignored\n");
	    with_color = 0;
	}
	else if ((A->type != MATRIX_TYPE_REAL) && !x) {
	    fprintf (stderr, "Warning: input has no coordinates, -c flag ignored\n");
	    with_color = 0;
	}
    }

    if (SparseMatrix_known_undirected(A)) {
	g = agopen("G", Agundirected, (Agdisc_t *) 0);
    } else {
	g = agopen("G", Agdirected, (Agdisc_t *) 0);
    }

    if (with_val) {
	sym = agattr(g, AGEDGE, "len", "1");
    }
    agxbinit (&xb, BUFS, string);
    if (with_label) {
	agxbput (&xb, name);
	agxbput (&xb, ". ");
	sprintf(buf, "%d", A->m);
	agxbput (&xb, buf);
	agxbput (&xb, " nodes, ");
	sprintf(buf, "%d", A->nz);
	agxbput (&xb, buf);
	agxbput (&xb, " edges.");
	agattr(g, AGRAPH, "label", agxbuse (&xb));
    }

    for (i = 0; i < A->m; i++) {
	sprintf(buf, "%d", i);
	n = agnode(g, buf, 1);
	agbindrec(n, "nodeinfo", sizeof(Agnodeinfo_t), TRUE);
	ND_id(n) = i;
	arr[i] = n;
    }

    if (with_color) {
	real maxdist = 0.;
	real mindist = 0.;
	int first = TRUE;

	sym2 = agattr(g, AGEDGE, "color", "");
	sym3 = agattr(g, AGEDGE, "wt", "");
	agattr(g, AGRAPH, "bgcolor", "black");
	color = N_NEW(A->nz, real);
	for (n = agfstnode(g); n; n = agnxtnode(g, n)) {
	    i = ND_id(n);
	    if (A->type != MATRIX_TYPE_REAL) {
		for (j = ia[i]; j < ia[i + 1]; j++) {
		    color[j] = distance(x, dim, i, ja[j]);
		    if (i != ja[j]) {
			if (first) {
			    mindist = color[j];
			    first = FALSE;
			} else {
			    mindist = MIN(mindist, color[j]);
			}
		    }
		    maxdist = MAX(color[j], maxdist);
		}
	    } else {
		for (j = ia[i]; j < ia[i + 1]; j++) {
		    if (val) color[j] = ABS(val[j]);
		    else color[j] = 1;
		    if (i != ja[j]) {
			if (first) {
			    mindist = color[j];
			    first = FALSE;
			} else {
			    mindist = MIN(mindist, color[j]);
			}
		    }
		    maxdist = MAX(color[j], maxdist);
		}
	    }
	}
	for (n = agfstnode(g); n; n = agnxtnode(g, n)) {
	    i = ND_id(n);
	    for (j = ia[i]; j < ia[i + 1]; j++) {
		color[j] =
		    ((color[j] - mindist) / MAX(maxdist - mindist,
						0.000001));
	    }
	}
    }

    i = 0;
    for (n = agfstnode(g); n; n = agnxtnode(g, n)) {
	i = ND_id(n);
	for (j = ia[i]; j < ia[i + 1]; j++) {
	    h = arr[ja[j]];
	    e = agedge(g, n, h, NULL, 1);
	    if (with_val && val) {
		sprintf(buf, "%f", val[j]);
		agxset(e, sym, buf);
	    }
	    if (with_color) {
		agxset (e, sym2, hue2rgb(.65 * color[j], cstring));
		sprintf(buf, "%f", color[j]);
		agxset(e, sym3, buf);
	    }
	}
    }

    agxbfree (&xb);
    FREE(color);
    FREE(arr);
    return g;
}

static char* useString = "Usage: %s [-uvcl] [-o file] matrix_market_filename\n\
  -u   - make graph undirected\n\
  -U i - treat non-square matrix as a bipartite graph\n\
         i = 0   never\n\
         i = 1   if pattern unsymmetric (default)\n\
         i = 2   if matrix unsymmetric\n\
         i = 3   always\n\
  -v   - assign len to edges\n\
  -c   - assign color and wt to edges\n\
  -l   - add label\n\
  -o <file> - output file \n";

static void usage(int eval)
{
    fprintf(stderr, useString, cmd);
    exit(eval);
}

static FILE *openF(char *fname, char *mode)
{
    FILE *f = fopen(fname, mode);
    if (!f) {
	fprintf(stderr, "Could not open %s for %s\n", fname,
		((*mode == 'r') ? "reading" : "writing"));
	exit(1);
    }
    return f;
}

typedef struct {
    FILE *inf;
    FILE *outf;
    char *infile;
    int undirected;
    int with_label;
    int with_color;
    int with_val;
  int bipartite;
} parms_t;

static void init(int argc, char **argv, parms_t * p)
{
    int c;

    cmd = argv[0];
    opterr = 0;
    while ((c = getopt(argc, argv, ":o:uvclU:")) != -1) {
	switch (c) {
	case 'o':
	    p->outf = openF(optarg, "w");
	    break;
	case 'l':
	    p->with_label = 1;
	    break;
	case 'u':
	    p->undirected = 1;
	    break;
	case 'v':
	    p->with_val = 1;
	    break;
	case 'c':
	    p->with_color = 1;
	    break;
	case 'U':{
	  int u;
	  if (sscanf(optarg,"%d",&u) <= 0 || u < 0 || u > BIPARTITE_ALWAYS) {
	    usage(1);
	  } else {
	    p->bipartite = u;
	  }
	  break;
	}
 	case ':':
	    fprintf(stderr, "%s: option -%c missing argument - ignored\n", cmd, optopt);
	    break;
 	case '?':
	    if (optopt == '?')
		usage(0);
	    else
		fprintf(stderr,
			"%s: option -%c unrecognized - ignored\n", cmd,
			optopt);
	    break;
	}
    }
    argv += optind;
    argc -= optind;

    if (argc > 0) {
	p->infile = argv[0];
	p->inf = openF(argv[0], "r");
    }
}

int main(int argc, char *argv[])
{
    Agraph_t *g = 0;
    SparseMatrix A = NULL;
    int dim=0;
    parms_t pv;

    /* ======================= set parameters ==================== */
    pv.inf = stdin;
    pv.outf = stdout;
    pv.infile = "stdin";
    pv.undirected = 0;
    pv.with_label = 0;
    pv.with_color = 0;
    pv.with_val = 0;
    pv.bipartite = BIPARTITE_PATTERN_UNSYM;
    init(argc, argv, &pv);

    /* ======================= read graph ==================== */

    A = SparseMatrix_import_matrix_market(pv.inf, FORMAT_CSR);
    if (!A) {
	fprintf (stderr, "Unable to read input file \"%s\"\n", pv.infile); 
	usage(1);
    }

    A = SparseMatrix_to_square_matrix(A, pv.bipartite);

    if (!A) {
	fprintf(stderr, "cannot import from file %s\n", pv.infile);
	exit(1);
    }

    if (pv.undirected) {
	SparseMatrix B;
	B = SparseMatrix_make_undirected(A);
	SparseMatrix_delete(A);
	A = B;
    }
    g = makeDotGraph(A, pv.infile, dim, NULL, pv.with_color, pv.with_label, pv.with_val);

    agwrite(g, pv.outf);

    return 0;
}

